File: libs/shared-types/src/zod/split-schemas.js
const fs = require('fs');
const path = require('path');

// Leer el archivo legacy
const content = fs.readFileSync('index.legacy.ts', 'utf-8');

// Definir los dominios y sus patrones
const domains = {
  'auth': {
    patterns: ['User', 'Role'],
    enums: ['RoleSchema', 'RoleType']
  },
  'messaging': {
    patterns: ['Message'],
    enums: []
  },
  'media': {
    patterns: ['Media'],
    enums: []
  },
  'profiles': {
    patterns: ['Profile', 'PortfolioItem'],
    enums: []
  },
  'companies': {
    patterns: ['Company', 'Branch'],
    enums: []
  },
  'services': {
    patterns: ['Service', 'Category', 'ServiceUnit', 'ServiceMetadata'],
    enums: []
  },
  'bookings': {
    patterns: ['Booking', 'BookingDetails', 'ServiceSlot', 'BookingStatus', 'SlotStatus'],
    enums: ['BookingStatusSchema', 'BookingStatusType', 'SlotStatusSchema', 'SlotStatusType']
  },
  'payments': {
    patterns: ['Payment'],
    enums: []
  }
};

// Función para extraer esquemas de un modelo
function extractSchemasForPattern(content, pattern) {
  const schemas = [];
  const lines = content.split('\n');
  
  let i = 0;
  while (i < lines.length) {
    const line = lines[i];
    
    // Buscar exports que contengan el patrón
    if (line.includes(`export const ${pattern}`) || line.includes(`export type ${pattern}`)) {
      const schemaLines = [line];
      let braceCount = (line.match(/{/g) || []).length - (line.match(/}/g) || []).length;
      i++;
      
      // Si es un tipo simple, solo tomar esa línea
      if (line.includes('export type') && line.includes('=') && braceCount === 0) {
        schemas.push(schemaLines.join('\n'));
        continue;
      }
      
      // Capturar el resto del esquema
      while (i < lines.length && (braceCount > 0 || !schemaLines[schemaLines.length - 1].trim().match(/[);}\]]$/))) {
        schemaLines.push(lines[i]);
        braceCount += (lines[i].match(/{/g) || []).length - (lines[i].match(/}/g) || []).length;
        i++;
        
        // Romper si encontramos una línea vacía después de cerrar
        if (braceCount === 0 && schemaLines[schemaLines.length - 1].trim() === '') {
          break;
        }
      }
      
      schemas.push(schemaLines.join('\n'));
    } else {
      i++;
    }
  }
  
  return schemas;
}

console.log("Extrayendo esquemas por dominio...\n");

// Crear directorios y archivos
for (const [domain, config] of Object.entries(domains)) {
  const domainDir = domain;
  if (!fs.existsSync(domainDir)) {
    fs.mkdirSync(domainDir, { recursive: true });
  }
  
  const allSchemas = [];
  
  // Extraer esquemas para cada patrón
  for (const pattern of config.patterns) {
    const schemas = extractSchemasForPattern(content, pattern);
    allSchemas.push(...schemas);
  }
  
  // Extraer enums específicos
  for (const enumName of config.enums) {
    const schemas = extractSchemasForPattern(content, enumName);
    allSchemas.push(...schemas);
  }
  
  if (allSchemas.length > 0) {
    // Crear archivo del dominio
    const domainFile = path.join(domainDir, 'index.ts');
    
    const fileContent = `import { z } from 'zod';
import { Prisma } from '@prisma/client';
import Decimal from 'decimal.js';
import {
  DecimalJsLikeSchema,
  isValidDecimalInput,
  JsonValueSchema,
  InputJsonValueSchema,
  NullableJsonNullValueInputSchema,
  JsonNullValueInputSchema,
  JsonNullValueFilterSchema,
} from '../common';

/////////////////////////////////////////
// ${domain.toUpperCase()} DOMAIN
/////////////////////////////////////////

${allSchemas.join('\n\n')}
`;
    
    fs.writeFileSync(domainFile, fileContent, 'utf-8');
    console.log(`✓ Creado ${domainFile} con ${allSchemas.length} esquemas`);
  }
}

console.log("\n✓ Extracción completada!");

File: libs/shared-types/src/zod/split-schemas.py
import re
import os

# Leer el archivo legacy
with open('index.legacy.ts', 'r', encoding='utf-8') as f:
    content = f.read()

# Definir los dominios y sus patrones
domains = {
    'auth': {
        'patterns': ['User', 'Role'],
        'enums': ['RoleSchema', 'RoleType']
    },
    'messaging': {
        'patterns': ['Message'],
        'enums': []
    },
    'media': {
        'patterns': ['Media'],
        'enums': []
    },
    'profiles': {
        'patterns': ['Profile', 'PortfolioItem'],
        'enums': []
    },
    'companies': {
        'patterns': ['Company', 'Branch'],
        'enums': []
    },
    'services': {
        'patterns': ['Service', 'Category', 'ServiceUnit', 'ServiceMetadata'],
        'enums': []
    },
    'bookings': {
        'patterns': ['Booking', 'BookingDetails', 'ServiceSlot', 'BookingStatus', 'SlotStatus'],
        'enums': ['BookingStatusSchema', 'BookingStatusType', 'SlotStatusSchema', 'SlotStatusType']
    },
    'payments': {
        'patterns': ['Payment'],
        'enums': []
    }
}

# Función para extraer esquemas de un modelo
def extract_schemas_for_pattern(content, pattern):
    schemas = []
    lines = content.split('\n')
    
    i = 0
    while i < len(lines):
        line = lines[i]
        
        # Buscar exports que contengan el patrón
        if f'export const {pattern}' in line or f'export type {pattern}' in line:
            schema_lines = [line]
            brace_count = line.count('{') - line.count('}')
            i += 1
            
            # Si es un tipo simple, solo tomar esa línea
            if 'export type' in line and '=' in line and brace_count == 0:
                schemas.append('\n'.join(schema_lines))
                continue
            
            # Capturar el resto del esquema
            while i < len(lines) and (brace_count > 0 or not schema_lines[-1].strip().endswith((')', '}', ';'))):
                schema_lines.append(lines[i])
                brace_count += lines[i].count('{') - lines[i].count('}')
                i += 1
                
                # Romper si encontramos una línea vacía después de cerrar
                if brace_count == 0 and schema_lines[-1].strip() == '':
                    break
            
            schemas.append('\n'.join(schema_lines))
        else:
            i += 1
    
    return schemas

print("Extrayendo esquemas por dominio...")

# Crear directorios y archivos
for domain, config in domains.items():
    domain_dir = domain
    if not os.path.exists(domain_dir):
        os.makedirs(domain_dir)
    
    all_schemas = []
    
    # Extraer esquemas para cada patrón
    for pattern in config['patterns']:
        schemas = extract_schemas_for_pattern(content, pattern)
        all_schemas.extend(schemas)
    
    # Extraer enums específicos
    for enum in config['enums']:
        schemas = extract_schemas_for_pattern(content, enum)
        all_schemas.extend(schemas)
    
    if all_schemas:
        # Crear archivo del dominio
        domain_file = f"{domain_dir}/index.ts"
        
        with open(domain_file, 'w', encoding='utf-8') as f:
            f.write("import { z } from 'zod';\n")
            f.write("import { Prisma } from '@prisma/client';\n")
            f.write("import Decimal from 'decimal.js';\n")
            f.write("import {\n")
            f.write("  DecimalJsLikeSchema,\n")
            f.write("  isValidDecimalInput,\n")
            f.write("  JsonValueSchema,\n")
            f.write("  InputJsonValueSchema,\n")
            f.write("  NullableJsonNullValueInputSchema,\n")
            f.write("  JsonNullValueInputSchema,\n")
            f.write("  JsonNullValueFilterSchema,\n")
            f.write("} from '../common';\n\n")
            f.write(f"/////////////////////////////////////////\n")
            f.write(f"// {domain.upper()} DOMAIN\n")
            f.write(f"/////////////////////////////////////////\n\n")
            f.write('\n\n'.join(all_schemas))
        
        print(f"✓ Creado {domain_file} con {len(all_schemas)} esquemas")

print("\n✓ Extracción completada!")
